File: algorithms/java/src/test/java/leetcode/LC372SuperPow.java
package leetcode;

import org.junit.jupiter.api.Test;


import java.util.LinkedList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/super-pow/
 * #medium
 */
public class LC372SuperPow {
    public class Solution {
        private static final int MODULO_1337 = 1337;

        private List<Integer> findPowerLoop(int value) {
            List<Integer> modulos = new LinkedList<>();
            boolean[] lut = new boolean[MODULO_1337];
            int modulo = value;
            while (!lut[modulo]) {
                lut[modulo] = true;
                modulos.add(modulo);
                modulo = (modulo * value) % MODULO_1337;
            }
            return modulos;
        }

        private int modulo(int[] dividend, int divisor) {
            int length = dividend.length;
            int modulo = 0;
            for (int i = 0; i < length; i++) {
                modulo = (modulo * 10 + dividend[i]) % divisor;
            }
            return modulo;
        }

        public int superPow(int a, int[] b) {
            // Assume: a = (1337 * n + m) where 0 <= m < 1337
            // Then: a^b mod 1337 = (1337 * n + m)^n mod 1337 == m^b mod 1337
            // This multiplication will cycle through certain "digits" of base 1337
            // You can search for the loop by iterating
            int m = a % MODULO_1337;
            List<Integer> modulos = findPowerLoop(m);
            // Get rid of loops from the power
            int length = modulos.size();
            int index = modulo(b, length);
            // Look-up the power modulo
            index = (index - 1 + length) % length;
            return modulos.get(index);
        }
    }

    @Test
    public void test_2_3() throws Exception {
        assertEquals(8, new Solution().superPow(2, new int[]{3}));
    }

    @Test
    public void test_2_1_0() throws Exception {
        assertEquals(1024, new Solution().superPow(2, new int[]{1, 0}));
    }
}

File: algorithms/code/leetcode/lc372_super_pow/lc372_super_pow.py
# https://leetcode.com/problems/super-pow/

import unittest


class Solution:
    __MODULO_1337 = 1337

    def __findPowerLoop(self, value: int) -> list[int]:
        modulos = []
        lut = [False] * self.__MODULO_1337
        modulo = value
        while not lut[modulo]:
            lut[modulo] = True
            modulos.append(modulo)
            modulo = (modulo * value) % self.__MODULO_1337
        return modulos

    def __modulo(self, dividend: list[int], divisor: int) -> int:
        length = len(dividend)
        modulo = 0
        for i in range(length):
            modulo = (modulo * 10 + dividend[i]) % divisor
        return modulo

    def superPow(self, a: int, b: list[int]) -> int:
        # Assume: a = (1337 * n + m) where 0 <= m < 1337
        # Then: a^b mod 1337 = (1337 * n + m)^n mod 1337 == m^b mod 1337
        # This multiplication will cycle through certain 'digits' of base 1337
        # You can search for the loop by iterating
        m = a % self.__MODULO_1337
        modulos = self.__findPowerLoop(m)
        # Get rid of loops from the power
        length = len(modulos)
        index = self.__modulo(b, length)
        # Look-up the power modulo
        index = (index - 1 + length) % length
        return modulos[index]


class TestCode(unittest.TestCase):
    def test_2_3(self) -> None:
        assert Solution().superPow(2, [3]) == 8

    def test_2_1_0(self) -> None:
        assert Solution().superPow(2, [1, 0]) == 1024
